proof/refine/AARCH64/Invariants_H.thy

```diff
@@ -883,31 +904,62 @@ where
+definition valid_bitmapQ :: "kernel_state \<Rightarrow> bool" where
+  "valid_bitmapQ \<equiv> \<lambda>s. \<forall>d p. bitmapQ d p s \<longleftrightarrow> \<not> tcbQueueEmpty (ksReadyQueues s (d,p))"
 
-definition
-  valid_queues :: "kernel_state \<Rightarrow> bool"
-where
- "valid_queues \<equiv> \<lambda>s. valid_queues_no_bitmap s \<and> valid_bitmapQ s \<and>
-                     bitmapQ_no_L2_orphans s \<and> bitmapQ_no_L1_orphans s"
+definition valid_bitmaps :: "kernel_state \<Rightarrow> bool" where
+  "valid_bitmaps \<equiv> \<lambda>s. valid_bitmapQ s \<and> bitmapQ_no_L2_orphans s \<and> bitmapQ_no_L1_orphans s"
 
-definition
-  (* when a thread gets added to / removed from a queue, but before bitmap updated *)
-  valid_bitmapQ_except :: "domain \<Rightarrow> priority \<Rightarrow> kernel_state \<Rightarrow> bool"
-where
+lemma valid_bitmaps_valid_bitmapQ[elim!]:
+  "valid_bitmaps s \<Longrightarrow> valid_bitmapQ s"
+  by (simp add: valid_bitmaps_def)
+
+lemma valid_bitmaps_bitmapQ_no_L2_orphans[elim!]:
+  "valid_bitmaps s \<Longrightarrow> bitmapQ_no_L2_orphans s"
+  by (simp add: valid_bitmaps_def)
+
+lemma valid_bitmaps_bitmapQ_no_L1_orphans[elim!]:
+  "valid_bitmaps s \<Longrightarrow> bitmapQ_no_L1_orphans s"
+  by (simp add: valid_bitmaps_def)
+
+lemma valid_bitmaps_lift:
+  assumes prq: "\<And>P. f \<lbrace>\<lambda>s. P (ksReadyQueues s)\<rbrace>"
+  assumes prqL1: "\<And>P. f \<lbrace>\<lambda>s. P (ksReadyQueuesL1Bitmap s)\<rbrace>"
+  assumes prqL2: "\<And>P. f \<lbrace>\<lambda>s. P (ksReadyQueuesL2Bitmap s)\<rbrace>"
+  shows "f \<lbrace>valid_bitmaps\<rbrace>"
+  unfolding valid_bitmaps_def valid_bitmapQ_def bitmapQ_def
+            bitmapQ_no_L1_orphans_def bitmapQ_no_L2_orphans_def
+  apply (rule hoare_pre)
+   apply (wps assms | wpsimp)+
+  done
+
+(* when a thread gets added to / removed from a queue, but before bitmap updated *)
+definition valid_bitmapQ_except :: "domain \<Rightarrow> priority \<Rightarrow> kernel_state \<Rightarrow> bool" where
   "valid_bitmapQ_except d' p' \<equiv> \<lambda>s.
-    (\<forall>d p. (d \<noteq> d' \<or> p \<noteq> p') \<longrightarrow> (bitmapQ d p s \<longleftrightarrow> ksReadyQueues s (d,p) \<noteq> []))"
+    \<forall>d p. (d \<noteq> d' \<or> p \<noteq> p') \<longrightarrow> (bitmapQ d p s \<longleftrightarrow> \<not> tcbQueueEmpty (ksReadyQueues s (d,p)))"
 
 lemmas bitmapQ_defs = valid_bitmapQ_def valid_bitmapQ_except_def bitmapQ_def
                        bitmapQ_no_L2_orphans_def bitmapQ_no_L1_orphans_def
 
-definition
-  valid_queues' :: "kernel_state \<Rightarrow> bool"
-where
- "valid_queues' \<equiv> \<lambda>s. \<forall>d p t. obj_at' (inQ d p) t s \<longrightarrow> t \<in> set (ksReadyQueues s (d, p))"
+\<comment> \<open>
+  The tcbSchedPrev and tcbSchedNext fields of a TCB are used only to indicate membership in
+  one of the ready queues. \<close>
```

Comment:
probably copied from MCS, but I think agreed style is to not put `― ‹` on its own line

Reply:
Oh, is it? I thought for multiline comments we generally did.

Reply:
Hmm, the style guide says to do `text ‹` on its own line, and I guess I was then extrapolating from that to `― ‹` as well.
https://github.com/seL4/l4v/blob/master/docs/Style.thy#L85

Reply:
I guess needs discussion and commitment into style guide, can take that out of this PR, @lsf can weigh in as well

Reply:
I guess the tension there is between what we do for `(*` and what we do for `text ‹`. `― ‹` is in between. I also hesitated at that one, tried it out on the same line, didn't like it either, and then left it for the same reason as Corey. Agreed that we should make this an issue for the style guide. (Maybe we should just not do multi-line `―` comments and use either `text` or `(* .. *)`)

Reply:
I think Ed used the style I use here quite a lot when we were doing MCS Refine, and I just picked up on that. It seems from the style guide that `text` is favoured, so I could change it to that if that's best.

Reply:
`text` won't work inside outer commands, only between them, so I think leave it for now, we don't seem to have a coherent policy

Reply:
+1 for issue for the style guide